ibis/backends/sql/rewrites.py

```diff
@@ -244,6 +255,22 @@ def merge_select_select(_, **kwargs):
     if _.parent.find_below(blocking, filter=ops.Value):
         return _
 
+    if not _.distinct and _.parent.distinct:
+        # SELECT ... FROM (SELECT DISTINCT ...)
+        if all(isinstance(v, ops.Field) for v in _.selections.values()):
+            # Can fuse a trivial subselection after a distinct, like
+            #   SELECT a, b FROM (SELECT DISTINCT ...) -> SELECT DISTINCT a, b, ...
+            distinct = True
+        else:
+            # Other cases aren't safe to fuse as they might introduce non-distinct
+            # columns where applying a distinct would change the result.
+            return _
+    else:
+        # SELECT DISTINCT ... FROM (SELECT DISTINCT ...) -> SELECT DISTINCT ...
+        # SELECT DISTINCT ... FROM (SELECT ...) -> SELECT DISTINCT ...
```

Comment:
I think this is only valid if the `distinct`-ed columns are guaranteed to be a subset-of-or-equal-to the parent columns. Pretty sure we don't have a way to construct anything else, but just wanted to call it out here.

Reply:
Are you saying this is only valid if the selection in the outer column is trivial (e.g. only column names)? If so, I don't see why that limitation is required, the following queries should be identical:

```sql
SELECT DISTINCT a, c % 2 AS d FROM (SELECT a, b + 1 AS c, FROM test);
SELECT DISTINCT a, (b + 1) % 2 AS d FROM test;
```

If I'm understanding correctly that you disagree, do you have a failing example?

Reply:
I'm asking whether we even have a way to construct `distinct` operations that are not always effectively `DISTINCT *`.

Reply:
Ah, yeah. There's no way to write those directly, but as select nodes are joined in `merge_select_select` they can come up. I believe I've covered all the possibilities with tests now.